### ISSUE TITLE
[BUG] Unable to load cookies [Changes on Udemy End]

### ISSUE THREAD

**Issue by parasguglani1** [aapatre/Automatic-Udemy-Course-Enroller-GET-PAID-UDEMY-COURSES-for-FREE#313]
**Describe the bug**
Unable to load cookies. Tried different VPNs  
one is giving
Exception in redeem courses: Detected a Cloudflare version 2 Captcha challenge, This feature is not available in the opensource (free) version.
 and other is
Exception in redeem courses: 'NoneType' object is not subscriptable
**To Reproduce**

execute udemy_controller --debug

**Expected behavior**
Should load the data from file

**Screenshots**
![image](https://user-images.githubusercontent.com/73358562/131625139-aec5dabd-3f66-4d98-8cfe-303ecded4c60.png)

**Desktop (please complete the following information):**
OS: Windows 10
Browser: Microsoft Edge Version 92.0.902.84 (Official build) (64-bit)
Version: release V 3.1.0
VPN: Proton
**Additional context**
I created new account with VPN because of issue #312 and it was working and got enrolled in 148 courses but today I am getting these two errors.

**Comment by fakeid30** [comment]
@cullzie I can't reproduce this. can you?

**Comment by Pos7alDude** [comment]
I got the same issue ('NoneType' object is not subscriptable) on Manjaro Linux. Yesterday it worked perfect. Deleted and re-created settings.yaml and cloned the repository again but doesn't solve the issue. I dit not run any updates after last use of the tool.

**Comment by ninjatall12** [comment]
i have gotten the same issue i think udemy did an update that requires people who use user-based logins to solve captcha i think the best way to solve this in the future is to use a cookie based auth. unfortunately im not familiar with web scraping so i can't help here.

**Comment by ninjatall12** [comment]
this is present with all udemy scrapers so its a udemy thing.

**Comment by cullzie** [comment]
@fakeid30 I am seeing something similar at the moment. Udemy have changed something on their side

**Comment by fakeid30** [comment]
ok now i can reproduce, also getting a lot of comments on social media.

**Comment by cullzie** [comment]
So there are 2 errors which show up here but they are both linked to the captcha appearing:

`Cloudflare version 2 Captcha challenge`
and
`Exception in redeem courses: 'NoneType' object is not subscriptable`

The fix isn't straightforward as we need a way to bypass the captcha which isn't currently a free service anywhere.
Unless there is another way to get around the captcha the script may no longer be able to enroll in udemy courses :disappointed: 
If anyone has suggestions or other possible solutions please post them here.

**Comment by mayur-chavhan** [comment]
There is an update for python module name cloudscraper, i'm not expert on python web scraping yet but module solution i think somewhere tricky to solve.

![image](https://user-images.githubusercontent.com/15307733/131823035-6ae951e7-6487-4fe5-8dd0-7220f96cad61.png)


https://pythonrepo.com/repo/VeNoMouS-cloudscraper-python-web-crawling

**Comment by Pos7alDude** [comment]
The error `Cloudflare version 2 Captcha challenge` is not new to me. It was always solved by running the script again until the 'NoneType' error occured.

**Comment by owlg** [comment]
@Pos7alDude running the script several times, results in either `Cloudflare version 2 captcha challenge` or `exception in redeem...` as described by cullzie, seems like it is not a solvable problem at the moment.

**Comment by cullzie** [comment]
So it looks like having it login in isn't the problem. The issue is when we try to access any of the endpoints with the generated cookie. I've tried with the curl request generated from the chrome dev tools and still hit the same issue.

**Comment by Pos7alDude** [comment]
@owlg Exactly what you write is currently the case. I wanted to clarify with my post that only the "captcha" message has been appearing from time to time for quite some time before.

**Comment by hoeni** [comment]
@Pos7alDude I hope you're right with the cookie-related error. But often big sites roll out new features (like a scraping protection) in several steps (blue/green deployment). So for a certain time only some servers have the change, e.g. to make performance or compatibility tests before it gets rolled out 100% in the end.

**Comment by ninjatall12** [comment]
![image](https://user-images.githubusercontent.com/58861295/132001322-7f59be79-d271-4f48-9385-4013bed19b71.png)
This is what you get

**Comment by cullzie** [comment]
So no requests outside the browser seem to work at the moment.
Previously we couldn't login on the browser hence we had to switch to REST requests.

We might need to look at a hybrid approach where we login via REST (which still seems to work) and then load those cookies into the browser. Not sure if this will work but I don't see another way at the moment.
Ideally we'd go fully REST or fully browser based but the net keeps closing in on what is possible as they reduce accessibility

**Comment by cullzie** [comment]
Has the script started working for anyone else? Its working for me now with no modifications :confused: 
Possibly the changes have been rolled back or there was some issue with the release causing problems which has been fixed.

**Comment by VasylTyshchuk** [comment]
Yes! It's also working for me now

**Comment by parasguglani1** [comment]
> Has the script started working for anyone else? Its working for me now with no modifications 😕
> Possibly the changes have been rolled back or there was some issue with the release causing problems which has been fixed.

working for me as well without VPN.

**Comment by fakeid30** [comment]
@cullzie not working for me, I get the captcha and another error (related to captcha, just a huge dump of broken ref) on a cyclic basis. Weird.

**Comment by cullzie** [comment]
@fakeid30 Looks like its regional deployments then. Not sure if they are rolling back or fixing issues at the moment though. We'll have to wait a bit and see if this gets to everyone

**Comment by fakeid30** [comment]
@cullzie false negative from my part. I was experimenting with the idea of hybrid login in local and today I ran the local branch instead of the develop branch from here. 

The dev branch is working for me too. 

But currently, if you input wrong password, it shows the captcha error. Just a small QoL.

**Comment by hoeni** [comment]
For me, the master branch wasn't working, but the dev branch worked (regular, German IP). It took a strange, long time for `Loading existing course details`, though, but went on.

**Comment by ninjatall12** [comment]
Afaik its returned bat to normal now but who knows if it will stay like this for long.

**Comment by parasguglani1** [comment]
it was working for me till yesterday but getting the same Cloudflare error today.

**Comment by ninjatall12** [comment]
> it was working for me till yesterday but getting the same Cloudflare error today.

hmm the update might depend on the region I reside in the UK I'm assuming the Cloudflare error can only happen once a day so try to use it at the same time you used it yesterday you should be Cloudflare unblocked.

**Comment by Pos7alDude** [comment]
The Cloudfare captcha error appears very often here (Germany) but, as I wrote above, it doesn't matter becaus you can simply re-run the script (sometimes you need three attempts) to get it working again. For me it's okay.

**Comment by cullzie** [comment]
Cloudflare captcha is a separate issue to the Udemy apis not working. Generally its because you've made too many requests within a certain time frame or you might be using a vpn that has a know vpn ip range to cloudflare.

For me the script is working at the moment

@Pos7alDude Would it make sense to have a bigger retry amount for udemy login? At the moment we re-try once. We could increase this which would save having to re-run the script multiple times. Not sure if it would work but if its something a lot of people are seeing it might be worth a try

**Comment by Pos7alDude** [comment]
@cullzie Maybe that would make sense for those who argue about the error message. But I'm fine as long as it works.

**Comment by Pos7alDude** [comment]
Now the "NoneType" error is occuring again since yesterday. No chance to run the script, tried multiple times.

**Comment by cullzie** [comment]
I think this is actually the 403 error. If you update to the latest release you should no longer have the NoneType error.

I have started seeing this again today and as far as I can see it isn't a straightforward fix.
Even curl commands with all info from the browser is hitting issues

`403 Client Error: Forbidden for url: https://www.udemy.com/api-2.0/https://www.udemy.com/api-2.0/users/me/subscribed-courses/?fields%5Bcourse%5D=title,completion_ratio&is_archived=false&ordering=-last_accessed&page=1&page_size=100
`

**Comment by Pos7alDude** [comment]
I'm using release 3.2.0. For testing I created new settings with only one option (language, category):
`udemy:
  categories: [Development]
  email: null
  languages: [English]
  password: null
  zipcode: '53773'`

But I still only get this output:
`
2021-10-15 08:12:36,003 - udemy_enroller - INFO - settings : Saved your settings in /home/seyffardtm/.udemy_enroller/settings.yaml
2021-10-15 08:12:36,004 - udemy_enroller - INFO - settings : Your email has not been saved to settings.
2021-10-15 08:12:36,004 - udemy_enroller - INFO - settings : Your password has not been saved to settings.
2021-10-15 08:12:36,004 - udemy_enroller - INFO - settings : You will be prompted to enter your email/password again when the cookie expires
2021-10-15 08:12:36,048 - udemy_enroller - INFO - udemy : Loading cookies from file
2021-10-15 08:12:36,128 - udemy_enroller - ERROR - runner : Exception in redeem courses: Detected a Cloudflare version 2 Captcha challenge, This feature is not available in the opensource (free) version.
2021-10-15 08:12:43,369 - udemy_enroller - INFO - settings : Loading existing settings
2021-10-15 08:12:43,382 - udemy_enroller - INFO - udemy : Loading cookies from file
2021-10-15 08:12:59,437 - udemy_enroller - INFO - udemy : Caching cookies for future use
2021-10-15 08:12:59,437 - udemy_enroller - INFO - udemy : Loading existing course details
2021-10-15 08:12:59,508 - udemy_enroller - INFO - udemy : Retrying login
2021-10-15 08:12:59,508 - udemy_enroller - INFO - udemy : Deleting cookies
2021-10-15 08:12:59,508 - udemy_enroller - INFO - udemy : Loading cookies from file
2021-10-15 08:13:00,667 - udemy_enroller - ERROR - runner : Exception in redeem courses: 'NoneType' object is not subscriptable
`

**Comment by Pos7alDude** [comment]
Now I tried several things. If I run `udemy_enroller` after `pip install udemy-enroller` I get either the cloudflare or the NoneType error. If I run `python3 /path-to-enroller/udemy_enroller.py` I get the 403 error. Any Ideas? A few days ago it worked with just running `undemy_enroller` (without VPN).
I'm from Germany, trying with and without VPN and using Manjaro Linux.

**Comment by cullzie** [comment]
@Pos7alDude You can run `pip install <package_name> -U` to upgrade a package
